interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,353 @@
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+)
+
+type componentType int
+
+// Component types in order from lowest to highest precedence.
+const (
+	compUnset componentType = iota
+	compGlobstar
+	compSeparatorDoublestarTerminal          // need to bundle separator and terminal marker with /**
+	compSeparatorDoublestarSeparatorTerminal // need to bundle separators and terminal marker with /**/
+	compSeparatorDoublestar
+	compTerminal // marker of end of pattern.
+	compSeparator
+	compAnySingle // ? has precedence over / so that /foo*?/ has precedence over /foo*/
+	compLiteral
+)
+
+type component struct {
+	compType componentType
+	compText string
+}
+
+func (c *component) String() string {
+	switch c.compType {
+	case compGlobstar:
+		return "*"
+	case compSeparatorDoublestarTerminal:
+		return "/**"
+	case compSeparatorDoublestarSeparatorTerminal:
+		return "/**/"
+	case compSeparatorDoublestar:
+		return "/**"
+	case compTerminal: // end of pattern
+		return ""
+	case compSeparator:
+		return "/"
+	case compAnySingle:
+		return "?"
+	case compLiteral:
+		return c.compText
+	}
+	return "###ERR_UNKNOWN_COMPONENT_TYPE###" // Should not occur
+}
+
+func (c *component) componentRegex() string {
+	switch c.compType {
+	case compGlobstar:
+		return `((?:[^/]|\\/)*)`
+	case compSeparatorDoublestarTerminal:
+		return `((?:/.+)?/?)`
+	case compSeparatorDoublestarSeparatorTerminal:
+		return `((?:/.+)?/)`
+	case compSeparatorDoublestar:
+		return `((?:/.+)?)`
+	case compTerminal:
+		return `(/?)`
+	case compSeparator:
+		return `(/)`
+	case compAnySingle:
+		return `([^/])` // does escaped '/' (e.g. `\\/`) count as one character?
+	case compLiteral:
+		return `(` + regexp.QuoteMeta(unescapeLiteral(c.compText)) + `)`
+	}
+	return `()`
+}
+
+var escapeFinder = regexp.MustCompile(`\\(.)`)
+
+func unescapeLiteral(literal string) string {
+	return escapeFinder.ReplaceAllString(literal, "${1}")
+}
+
+type PatternVariant struct {
+	variant    string
+	components []component
+	regex      *regexp.Regexp
+}
+
+func (v *PatternVariant) String() string {
+	return v.variant
+}
+
+// ParsePatternVariant parses a rendered variant string into a PatternVariant
+// whose precedence can be compared against others.
+func ParsePatternVariant(variant string) (*PatternVariant, error) {
```

Comment:
I would not heap-allocate the result as the value is small and w stash them into a slice of pointers. Nitpick.

Reply:
Good idea! How about when using the variants after they have been collected into a slice? Here's where they remain used with pointers:

```
me@hostname:~$ grep -r '\*PatternVariant'
variant.go:func (v *PatternVariant) String() string {
variant.go:func (v PatternVariant) Compare(other *PatternVariant, matchingPath string) (int, error) {
patterns.go:func HighestPrecedencePattern(patterns []*PatternVariant, matchingPath string) (*PatternVariant, error) {
```

Reply:
I made these all act on/take `PatternVariant` instead of `*PatternVariant`. And same for `component`. But not for `componentReader`, since it needs to mutate the receiver.